### PR TITLE
Add fish support and tests

### PR DESCRIPTION
**Pull request by flybonzai**
Addresses this issue: https://github.com/rogusdev/aws-cli-mfa/issues/6

### CODE REVIEW

File: bin/aws-cli-mfa-fish
Comment (flybonzai) [inline bin/aws-cli-mfa-fish:21]: The difference between heredocs and fish multiline strings was tricky, so we have to pre-process the python script to escape any quotes that conflict with the top level.

File: bin/aws-cli-mfa-fish
Comment (flybonzai) [inline bin/aws-cli-mfa-fish:158]: fish variables are all arrays, so to get multiline strings into a variable you need to call `string collect` on them.

File: build.sh
Comment (flybonzai) [inline build.sh]: Sourcing this file, even though it is on `$PATH`, does not work in fish (by design according to one of the committers in gitter).  To get around this one option would be to wrap it in a script that sources the file using its full path.

Comment (flybonzai) [reply build.sh]: That being said, I was able to get session tokens and make aws calls locally using the new fish scripts, I just had to use the full path for `aws-cli-mfa`.

File: tests/tests.fish
Comment (flybonzai) [inline tests/tests.fish]: I can get some of the tests working when I put them in the `.sh` tests script, but there are a lot of differences that would be tricky to work around.  For example, for fish instead of `&& env` it needed to be `; env`.  Separate test file is better than nothing.

File: src/mfa.fish
Comment (flybonzai) [inline src/mfa.fish]: `IS_SOURCED` is a hacky workaround for the fact that fish doesn't set `$_` the same when it is sourced from inside a command substitution.  Probably a better way to do this.

File: tests/tests.fish
Comment (flybonzai) [inline tests/tests.fish:5]: `test_program` so we don't shadow the built-in `test`

File: bin/aws-cli-mfa-fish
Comment (rogusdev) [inline bin/aws-cli-mfa-fish:17]: This is different from the others: they only quit at this point if it was sourced -- the reason for this is to support getting usage instructions without sourcing and without installing all requirements first

File: bin/aws-cli-mfa-fish
Comment (rogusdev) [reply bin/aws-cli-mfa-fish:21]: ~Why did you put the entire python code in this script? The entire point is to separate the python code and the shell scripts around that. This makes it so we would have to update the python code in 2 places now, which is definitely not ok -- and python tests cannot be run against this version of the code at all.~

File: bin/aws-cli-mfa-fish
Comment (rogusdev) [inline bin/aws-cli-mfa-fish:148]: Same as above with these exits: this is different in the other shells, on purpose, to only quit if sourced

File: bin/aws-cli-mfa-fish
Comment (flybonzai) [reply bin/aws-cli-mfa-fish:21]: The other files in `bin` also have the python code embedded.  This file was generated by the build command.  The confusion is probably around that I put the majority of my comments on this file and not the one in `src`.

File: bin/aws-cli-mfa-fish
Comment (rogusdev) [inline bin/aws-cli-mfa-fish]: Nit: git, and I, prefer trailing newlines

File: build.sh
Comment (rogusdev) [inline build.sh]: What is this extra sed here? Is this what you were referring to above about quotes?

File: bin/aws-cli-mfa-fish
Comment (rogusdev) [reply bin/aws-cli-mfa-fish:21]: Ohhhhhh I completely ignored that this is the bin version of the script!

File: tests/tests.fish
Comment (rogusdev) [inline tests/tests.fish:1]: So, the tests are explicitly written assuming that you have bash, and can trigger the other shells from there. As in, the tests are meant to all run from one file, that tests all envs. You created a separate file for fish tests, which is not the intent. Was there a reason this needed to happen?

File: tests/tests.fish
Comment (rogusdev) [reply tests/tests.fish]: Got it. Now I need to compare line by line that the tests are the same tho! Ha, will do later

File: tests/tests.fish
Comment (rogusdev) [inline tests/tests.fish]: Same nit about trailing newlines

**Comment by rogusdev** [comment]
Please explicitly reference the issue in your PR description, since you are resolving an issue ( #6 )

File: src/mfa.fish
Comment (rogusdev) [reply src/mfa.fish]: It doesn't set it the same, or not at all?

Comment (flybonzai) [reply src/mfa.fish]: That is a good question, I'll have to test that specifically.  If sourcing indirectly, like `fish -c 'source ...'` that test returns false.

File: build.sh
Comment (flybonzai) [reply build.sh]: Yes, this is to escape the `"` to not conflict with setting the variable.

File: tests/tests.fish
Comment (rogusdev) [inline tests/tests.fish]: It might seem unnecessary in context, but especially for consistency, as well as for potential future proofing, I would actually prefer that you retain the `TEST_SHELL` variable, just setting it to "fish" ofc. And I mention this here because these lines are missing that these are fish tests -- again, clear when you are deliberately running this script, but if this was a part of CI or whatever, it might not be as obvious at some point.

File: src/mfa.fish
Comment (rogusdev) [reply src/mfa.fish]: Add an echo line to see what gets set, and maybe you can still use whatever is there

File: tests/tests.fish
Comment (rogusdev) [inline tests/tests.fish]: This error message is incorrect, it should be `python3 could not be found`

Comment (rogusdev) [reply tests/tests.fish]: https://github.com/rogusdev/aws-cli-mfa/blob/642b9356122ba1fd2068133ac3d168a51f760f47/tests/tests.sh#L50

File: tests/tests.fish
Comment (rogusdev) [inline tests/tests.fish]: Same with these -- you are string replacing too much, your setup is different from the originals

File: bin/aws-cli-mfa-fish
Comment (flybonzai) [reply bin/aws-cli-mfa-fish:17]: I'm not a bash expert, but I'm a little confused here.  

Using this little test script, if I understood your intent correctly, sourcing it should exit out in the `if`, but calling it directly should print out that python couldn't be found, and then print Yoooo?  That wasn't the case in my testing, but I'm probably missing something here.

```bash
#!/usr/bin/env bash

([[ -n $ZSH_EVAL_CONTEXT && $ZSH_EVAL_CONTEXT =~ :file$ ]] ||
 [[ -n $KSH_VERSION && $(cd "$(dirname -- "$0")" &&
    printf '%s' "${PWD%/}/")$(basename -- "$0") != "${.sh.file}" ]] ||
 [[ -n $BASH_VERSION ]] && (return 0 2>/dev/null)) &&
    SOURCED=1 || SOURCED=0

if ! command -v python1 >/dev/null 2>&1; then
    echo "python3 could not be found!"
    [[ $SOURCED -eq 1 ]] && return 0 || exit 0
fi

echo "Yoooo"
```

File: tests/tests.fish
Comment (flybonzai) [reply tests/tests.fish]: Yeah this was a result of shortening the requirements check since it used the same variable we were looping through.  I can expand this back out to maintain consistency with the others.

File: bin/aws-cli-mfa-fish
Comment (rogusdev) [reply bin/aws-cli-mfa-fish:17]: Ah, turns out I misremembered: it is because `return` means something different in a bash script, depending on if it was sourced or not -- so maybe this does not apply to fish.

Test script I used:
```
#!/usr/bin/env bash

# https://stackoverflow.com/questions/2683279/how-to-detect-if-a-script-is-bein$
([[ -n $ZSH_EVAL_CONTEXT && $ZSH_EVAL_CONTEXT =~ :file$ ]] ||
 [[ -n $KSH_VERSION && $(cd "$(dirname -- "$0")" &&
    printf '%s' "${PWD%/}/")$(basename -- "$0") != "${.sh.file}" ]] ||
 [[ -n $BASH_VERSION ]] && (return 0 2>/dev/null)) &&
    SOURCED=1 || SOURCED=0

echo $SOURCED

if ! command -v aws1 >/dev/null 2>&1; then
    echo "aws could not be found!"
    [[ $SOURCED -eq 1 ]] && return 0 || exit 0
#    return 0
fi

echo "completed"
```

My experiments with it, first with the actual SOURCED check && return || exit, then with just return:
```
bash-3.2$ . test.sh
1
aws could not be found!
bash-3.2$ ./test.sh
0
aws could not be found!
bash-3.2$ nano test.sh # swap out commented line to enable just return 0
bash-3.2$ ./test.sh
0
aws could not be found!
./test.sh: line 15: return: can only `return' from a function or sourced script
completed
bash-3.2$ . test.sh
1
aws could not be found!
```

Comment (rogusdev) [reply bin/aws-cli-mfa-fish:17]: Indeed, adapting my example for fish shows that fish does not have this particular bash quirk:

Script (hardcoded SOURCED):
```
#!/usr/bin/env fish

set SOURCED 0

echo $SOURCED

if not command -v aws1 &>/dev/null
    echo "aws could not be found!"
#    [[ $SOURCED -eq 1 ]] && return 0 || exit 0
    exit 0
end

echo "completed"
```

Output:
```
chris.rogus@BZUSC02YT0FALVCH ~> . test.sh
1
aws could not be found!
chris.rogus@BZUSC02YT0FALVCH ~> ./test.sh
1
aws could not be found!
chris.rogus@BZUSC02YT0FALVCH ~> nano test.sh # change SOURCED to 0
chris.rogus@BZUSC02YT0FALVCH ~> ./test.sh
0
aws could not be found!
chris.rogus@BZUSC02YT0FALVCH ~> . test.sh
0
aws could not be found!
```

Comment (rogusdev) [reply bin/aws-cli-mfa-fish:17]: So, this is not a problem in fish -- resolving

File: bin/aws-cli-mfa-fish
Comment (rogusdev) [reply bin/aws-cli-mfa-fish:148]: As above, not a thing in fish -- resolved

File: tests/tests.fish
Comment (rogusdev) [reply tests/tests.fish]: Ah, if that's the case, I suppose I can live with this

File: tests/tests.fish
Comment (rogusdev) [reply tests/tests.fish]: As above, this is fine

File: src/mfa.fish
Comment (rogusdev) [reply src/mfa.fish]: This is actually broken already, based on basic testing (it does not work for `.`, only `source`):

Script:
```
#!/usr/bin/env fish

echo "$_"
```

Output:
```
$ . test.sh 
.
$ source test.sh
source
```

Comment (rogusdev) [reply src/mfa.fish]: It looks like you need to use `status current-command` instead of `$_` per https://fishshell.com/docs/current/language.html#envvar-_ which also resolves this problem in a reasonable way, it outputs `fish` when run like

```
$ echo (fish -c "source test.sh" | string collect)
fish
$ fish test.sh
fish
```

Which frankly makes a lot of sense and I wonder if that should be supported for bash/etc (i.e. running directly with `bash bash.sh` etc)

Comment (rogusdev) [reply src/mfa.fish]: So I think this specific line should instead be something like:

```
#!/usr/bin/env fish

set IS_SOURCED_COMMANDS . source fish
echo $IS_SOURCED_COMMANDS
set SOURCE (status current-command)
echo $SOURCE
if contains $SOURCE $IS_SOURCED_COMMANDS
  set SOURCED 1
else
  set SOURCED 0
end
echo $SOURCED
```

HOWEVER... this reveals that:
```
chris.rogus@BZUSC02YT0FALVCH ~> . test.sh
. source fish
.
1
chris.rogus@BZUSC02YT0FALVCH ~> source test.sh
. source fish
source
1
chris.rogus@BZUSC02YT0FALVCH ~> fish test.sh
. source fish
fish
1
chris.rogus@BZUSC02YT0FALVCH ~> ./test.sh
. source fish
fish
1
```
^^ note that we explicitly do not want that last one to work... Or do we? So, in bash, that last case definitely means that env vars will not be set, which is why sourcing is a hard requirement. But... perhaps fish setting env vars in a script will actually work outside, regardless of source-ing the file. So, maybe this requirement is no longer applicable for fish??

Comment (rogusdev) [reply src/mfa.fish]: Hmm, no, as it happens, fish does the same thing as bash: you need to source the file to get env vars. UNLESS you do `set -xU` which makes it available to all users shells across the machine. But this seems like a very dangerous (security risk) move, so I would avoid it. That means we do not want to include `fish` in the list of acceptable sources, and that means we need to set an env var before running the `fish -c ...` tests. But we can make things easier by doing something simpler (most definitely cheating) like: `if contains $SOURCE $SOURCES; or $SOURCED; set $SOURCED 1; end` something like that -- and just `set $SOURCED 1` in the test file

File: build.sh
Comment (rogusdev) [inline build.sh]: instead of the sed, perhaps you should use fish's `string escape`? https://fishshell.com/docs/current/cmds/string.html

File: src/mfa.fish
Comment (rogusdev) [reply src/mfa.fish]: In short, at this point, I think the only thing that needs to change here, for sure, is: accept `.` as well as `source`. That said, I suggest also using `status current-command` instead of `$_`. Past that, what you have here as your workaround for `fish -c ...` seems acceptable.

File: build.sh
Comment (flybonzai) [reply build.sh]: I like this idea.  I wasn't sure if it was a good idea to mix fish code in the same file as bash/zsh/etc, but these clearly are meant to be used individually.

File: src/mfa.fish
Comment (flybonzai) [reply src/mfa.fish]: Great find here, I've added what you suggested.  Also, digging how expressive it is to do things like see if a term is contained in a list.

File: build.sh
Comment (flybonzai) [reply build.sh]: `string escape` seems aimed at sanitizing fish script, urls and variables.  I ran a few tests, but wasn't able to get it to do the thing we need, `string replace --all '"' '\\"' < input_file` works great though.  I changed it in both the tests file and the `build.sh`

File: tests/tests.fish
Comment (flybonzai) [reply tests/tests.fish]: Fixed this.

File: build.sh
Comment (rogusdev) [reply build.sh]: Tbh, if the replacement for a quick and easily read sed statement is that, I might be inclined to keep the sed statement... But up to you if you prefer the new one

**Comment by rogusdev** [comment]
Please also update the README to mention `.fish_profile` and `.zshrc` (I forgot it before, clearly), same as the instructions for bash_profile.

And while you could also mention the `. (which aws-cli-mfa)` as the fish workaround for not recognizing new user bin apps as a global applications, ~but that just seems absolutely completely impossible, that there is no way to add new applications to run this way.~ OK, experimenting a bit, clearly this is just because fish doesn't think you should source files from there, very inconvenient. Do you have a link to the fish explanation on this?

**Comment by rogusdev** [comment]
Also, I am going to squash merge unless you do all the merging yourself, if you want. I don't need all the little step wise commits in history, for my tastes

**Comment by flybonzai** [comment]
> Also, I am going to squash merge unless you do all the merging yourself, if you want. I don't need all the little step wise commits in history, for my tastes

I'll make sure to squash them.

**Comment by flybonzai** [comment]
> Please also update the README to mention `.fish_profile` and `.zshrc` (I forgot it before, clearly), same as the instructions for bash_profile.
> 
> And while you could also mention the `. (which aws-cli-mfa)` as the fish workaround for not recognizing new user bin apps as a global applications, ~but that just seems absolutely completely impossible, that there is no way to add new applications to run this way.~ OK, experimenting a bit, clearly this is just because fish doesn't think you should source files from there, very inconvenient. Do you have a link to the fish explanation on this?

They don't really explain their reasoning here, but it is in the docs for source: https://fishshell.com/docs/current/cmds/source.html.  

`fish will search the working directory to resolve relative paths but will not search [PATH](https://fishshell.com/docs/current/language.html#envvar-PATH) .`

I asked one of the main devs in their gitter, and this was his response:
image

**Comment by flybonzai** [comment]
@rogusdev should be ready for final/near final review.  Updated README to include all things Fish, and I split out the fish build command into its own file since someone wanting to use this tool in fish (likely) won't care about building all the other shell files.

File: README.md
Comment (rogusdev) [inline README.md]: Does this also copy it to bin, like the above instructions do? If not, please include the copy instructions on the line as well

File: build.sh
Comment (rogusdev) [inline build.sh]: This file is running in bash, per the shebang. Please keep fish in here, but use the correct bash syntax.

File: build.fish
Comment (rogusdev) [inline build.fish]: Nit: please have a trailing newline

File: README.md
Comment (rogusdev) [inline README.md]: These readme changes are all great, thanks! (Just the one ask above)

File: build.sh
Comment (flybonzai) [reply build.sh]: Updated.  Curious why the extra hop in the build script into `/usr/local/bin/aws-cli-mfa` just to copy it into the local `bin` and then remove the copy in `/usr/local/bin`?  Since the README instructions have you copy the file out of the local `bin` back into the original location, should the commands in this script just build the local `bin` files?

File: README.md
Comment (flybonzai) [reply README.md]: Yes, it is the same command as is in the `build.sh` file.  As per the comment below, they actually all copy it into `/usr/local/bin`, but since the fish script only has the one command we don't need to do the extra `rm` and then copy it back.

File: README.md
Comment (flybonzai) [reply README.md]: Sounds good, addressed them and left a question and a comment

File: build.sh
Comment (rogusdev) [reply build.sh]: Hmm could you add a separate commit showing what you are suggesting cleaning up? It sounds intriguing but I'm not sure I follow

Comment (flybonzai) [reply build.sh]: Here is what I had in mind:
https://github.com/rogusdev/aws-cli-mfa/pull/9/commits/633100121064cdd5525f470c1fb1fdc994f75d4f

It effectively removes an extra copy since at the end of the script we are deleting `/usr/local/bin/aws-cli-mfa` anyway.  Also, it seems to have fixed a bug where `help='do not export/set AWS_PROFILE to the one generated by STS')` was duplicated for `--lifetime-duration` arg as well.

File: build.sh
Comment (rogusdev) [inline build.sh]: You should add the chmod +x for all of these

File: bin/aws-cli-mfa-ksh
Comment (rogusdev) [inline bin/aws-cli-mfa-ksh]: Oh, yeah, this was me -- I forgot to run build after changing this: https://github.com/rogusdev/aws-cli-mfa/commit/2ba000ad47ec686c5d6d68ff458fbc3f8aad2d84 -- so I can fix that separately from your PR in a bit

File: build.fish
Comment (rogusdev) [inline build.fish]: Yeah, I am pretty sure this is a result of me writing these instructions once, for the original bash, and then when I did it for other shells, I just copy-pasted and did not think about it hard enough. I support cleaning this up as you have suggested in build.sh, and doing it here as well. Please do add the chmod +x as well tho, as noted below

File: build.sh
Comment (flybonzai) [reply build.sh]: Ok, added.

File: README.md
Comment (rogusdev) [inline README.md]: Is there a reason for this extra newline?

File: build.fish
Comment (rogusdev) [reply build.fish]: Don't you want to change the steps here to match what you did in build.sh?

File: README.md
Comment (flybonzai) [reply README.md]: Nope, fixed.

File: build.fish
Comment (flybonzai) [reply build.fish]: We don't need the additional fish build script anymore, so I updated the README so that the user doesn't need to set a env variable to do the copy (which was what was breaking it for fish).  LMK if you don't like that change.

File: README.md
Comment (rogusdev) [reply README.md]: Are you sure it's fixed? Still showing up

File: build.sh
Comment (rogusdev) [inline build.sh:12]: This looks like a bad copy paste from fish. Will this file exist?

File: README.md
Comment (flybonzai) [reply README.md]: Rats, not sure how that snuck back in.  Now it is actually fixed :D

**Comment by flybonzai** [comment]
@rogusdev do we have any pending issues you want addressed before I squash again in preparation for merge?

File: bin/aws-cli-mfa-ksh
Comment (rogusdev) [reply bin/aws-cli-mfa-ksh]: You need to rebase/merge now -- I updated these correctly on main

**Review by rogusdev — APPROVED** [review]
Rebase and squash and it'll be all good for me, thanks!

**Comment by flybonzai** [comment]
@rogusdev squashed and ready.

**Comment by rogusdev** [comment]
You didn't squash, but I can do it for you

**Pull request merged**
